### ISSUE TITLE
Extension is no longer working after browser updates and manifest v3 (SOLVED).

### ISSUE THREAD

**Issue by XanSama** [rain-fighters/AudioPick#47]
[AudioPick_ManifestV3.zip](https://github.com/rain-fighters/AudioPick/files/13708745/AudioPick_ManifestV3.zip)

I don't feel like submitting a pull request but the attached is a rewritten version of the plugin that is confirmed working with the latest builds of Chrome and Edge (major version 120).

If the author wants they can update the repo/published plugin.
If not this can be extracted and loaded as an unpacked extension as long as developer mode is enabled in the browser.

**Comment by lrq3000** [comment]
Thank you for making this but please make a PR, not only this will help devs merge faster and with proper attribution, this will also help users ensure there is no rogue code as happend with other popular extensions.

**Comment by necropola** [comment]
Interesting!

**Comment by necropola** [comment]
@XanSama I've created a `manifest-v3` branch and made it the (temporary) default for the repository. Feel free to file a PR with your changes.

Cheers

**Comment by necropola** [comment]
@XanSama  Are you sure that the `attachShadow` hook actually works?
-  Do you have an example of a site which inserts their `<audio />` and/or `<video />` elements into a Shadow DOM and become accessible to **AudioPick's** `setSinkId` method by attaching the observer to their **Shadow DOM**, too?
- According to https://stackoverflow.com/questions/54954383/override-element-prototype-attachshadow-using-chrome-extension you would need to inject an additional script to hook `attachShadow` instead of hooking from the content script itself. Is this not required?

I think that the biggest challenge are sites which do not insert their audio/video elements into the DOM tree at all, e. g. **Spotify** which (according to Chrome Developer Tools / Heap Snaphot) has a **Detached Video Element**. Getting access to those would be really cool.

**Comment by XanSama** [comment]
[AudioPick-v0.3.6.zip](https://github.com/rain-fighters/AudioPick/files/13758516/AudioPick-v0.3.6.zip)

Good catch on the attachShadow; site I was testing with apparently doesn't always use them.
I've updated the resources example with a shadowRoot test and confirmed that this version is working with it.

Unfortunately this version required blanket host_permissions for <all_urls> as we need to be able to inject script into WORLD:MAIN to make the hooks work as intended.

I also added a hook for HTMLMediaElement.prototype.play and that resolves the issue with the hidden object in the test page.

Confirmed working on all tests, Spotify, new Reddit, and YouTube.

Still don't feel like making a PR so here's another attachment.
Don't worry about attribution.

**Comment by necropola** [comment]
@XanSama Nice!

I did not know about this **Hooking Trick** (inject a function into ***Element***`.prototype.`***method***) before I saw it in your code for attaching the observer to **ShadowRoots**.

And I already thought that hooking into a ***HTMLMediaElement*** method might be what we need for sites like **Spotify**.  Though my idea was to somehow hook into the constructor itself and remember those deatched elements for when we want to switch their `sinkId`.

Your approach (hooking the `play` method) is simpler (more elegant) and avoids **AudioPick** holding onto detached elements which actually should be garbage collected, but it requires the `play` method to be triggered after changing the assigned `sinkId`.

By the way, I'm perefctly happy with you providing your code/changes as attachments to this issue. I'll soon update the `manifest-v3` branch with a version based on your code. I'll just do some cosmetic changes and try to fix some errors, e. g. when the popup tries to send a meesage to a tab which does not have a listener yet.

I'm also currently working on a new version of the **Rain-Fighters/AudioPick** web site (using **hugo**) which I want to have proper documentation about how the new (`manifest-v3`) version of **AudioPick** works and where I got the code from.

As soon as the web site is ready and I've tested, understood and documented the v3 version of the extension well enough, I'll update the Chrome Web Store page, too. (Probably early 2024).

Which reminds me .... **Merry Christmas** and a **Happy New Year** to everyone and a **BIG THANK YOU to @XanSama** who made us this xmas present.

**Comment by necropola** [comment]
@XanSama Can you give me your GitHub `no-reply` email address so I can add a proper `Co-authored-by: ...` entry to the commit message for importing your changes? You'll find it under `Settings/Email/Keep my email address private` and it should look like this 
```
1234567+xansama@users.noreply.github.com
```
I basically need to know those leading digits ...

**Comment by XanSama** [comment]
Address is: "109573827+XanSama"...
[AudioPick-v0.3.7.zip](https://github.com/rain-fighters/AudioPick/files/13760632/AudioPick-v0.3.7.zip)

I'd initially not hooked the constructor because it doesn't seem to be reliably called for several ways that the objects can be created but your comment inspired me to take another look. I've updated it to hook Document.prototype.createElement in addition to the others as in theory that needs to be called to create an element outside of the DOM/shadowDOM. I've also adjusted the createElement() and play() hooks to create an eventListener per audio/video element that watches for the custom event "changeSinkId". This allows us to set the sinkId on all elements immediately rather than waiting for the next play(). Also added a dirty catch for that runtime error to suppress it; could probably actually handle the error but it doesn't seem to matter in my testing.

**Comment by necropola** [comment]
Hmm, does hooking `Document.prototype.createElement` also catch `hidden_audio = new Audio(...)`? I really like the idea of the custom event listener, but I also feel likely the codes is getting a bit too complex.

**Comment by XanSama** [comment]
Unfortunately not. The play() hook does address that though. Honestly we may be able to do without the createElement hook and just use the play() hook with the event listener for immediate sink changes.

**Comment by necropola** [comment]
yeah, lets not make things too complicated. I'll push your 0.3.6 version in a moment. Please use the updated version on GitHub as a base for you next  version (with the custom event listener added via the play() hook).

**Comment by necropola** [comment]
@XanSama pushed your `0.3.6` version  + some white space (4  spaces -> 1 tab) and EOL (CRLF -> LF) cleanup. 😸

**Comment by XanSama** [comment]
[AudioPick-v0.3.7.zip](https://github.com/rain-fighters/AudioPick/files/13760774/AudioPick-v0.3.7.zip)

Re-based and added the event trigger.
Left the createElement hook out for now. You can pull it from the last zip before this one if it ends up looking useful.

**Comment by necropola** [comment]
@XanSama hmm, looks like you attached the wrong (same as before) archive.

**Comment by XanSama** [comment]
Woops. Edited with correct version now.

**Comment by necropola** [comment]
@XanSama Imported and pushed `0.3.7`. Looks good!

I really like the `changeSinkId` event method for immediate sink switching on **Spotify** and the likes ... but it's very late/early over here in Germany now and I need to get some sleep ... Merry Christmas!

P. S.: I'm still wondering whether we need to attach to ShadowRoots or even need to observe the DOM tree with the event trigger in place ... hmm ... Good Night!

**Comment by XanSama** [comment]
Merry Christmas!
Hah. I thought I'd tested that and it hadn't worked but I've retested it like that and it seems like a Christmas miracle.

Tests OK on all examples, Reddit, and Spotify using only the play() hook.

It seems to test OK with the shadowRoot code disabled as well... theoretically if the site managed to load/create an element in the shadowDom that the user could directly interact with -before- our content script loaded it wouldn't catch that with the shadowRoot code disabled but I'm having a hard time imagining how one would even do that to test. Can tentatively nuke all of that.

Tentatively attaching 
[main.zip](https://github.com/rain-fighters/AudioPick/files/13760880/main.zip)
a stripped down main.js that only implements the hook/eventListener.

**Comment by necropola** [comment]
@XanSama  Almost!

The issue is that clicking `play ⏯️` on a visible `<audio/>` (or `<video/>`) element' does not trigger the **HTMLMediaElement's** `play()` method, hence it does not insert our `changeSinkId` listener.

To reproduce the issue use our example page and
- set the **domain** or **gbobal** sink to something other than **default**
- reload the page
- click the `Add new Visible` and/or `Add Visible in shadowRoot` button
- start the audio by clicking `play ⏯️` on the now visible element
- observe that it still uses the default audio device

I still think that the custom event listener is the way to go, i. e. the actual `sinkId` switching should always be handled by dispatching the `changeSinkId` event, but we use different methods to inject the listener.

#### (A) Visible/Attached HTMLMediaElements
These are either already in the DOM tree or are inserted later. The former we catch by traversing the tree once (-> `setAllSinks()` rename to `prepareAllSinks()`) and the latter we catch by using a **Mutation Observer** (we originally had one).

But instead of (or additionally to) changing the `sinkId` we insert our event listener for `changeSinkId` events. Note that elements found by the observer mighht already have the event listener added, if their `play()` method has been invoked while they were still detached. See (B)

#### (B) Detached HTMLMediaElements
Those are dynamically created at runtime and are usually not inserted into the DOM tree later. We assume that the only way to activate them while they are not visible is by calling their `play()` method, hence we insert our event listener by hooking into the **prototype's** `play()` method.

Note that for this approach it does not matter whether an element was attached to a **shadowRoot** or not. Or am I overlooking something?

What do you think?

P. S.: I havent commited and pushed the new `main.js` due to the above mentioned issue.

**Comment by XanSama** [comment]
Ahh, that was what I ran in to in my initial testing but missed it after I adjusted the play/pause buttons because I wasn't interacting directly with the media element.

Confirmed / reproduced / resolved.
Mutation observer and attachShadow hook are back in place to address the regression.

I've updated the code that sets the sinkId to set the sinkId and add a listener to an object on first pass, or leave it alone if it already has a listener. Objects generally all handle their own sinkId updates after the initial set.

I tweaked the injection (set it to all frames) and the content.js message handling (only register a handler in the top window).
The async error has also been resolved. I can't get it to pop any more errors but feel free to give it a kick.

SoundCloud appears to be using AudioContext objects for their streaming playback. I didn't know those existed until just now but I've implemented a hook of them as well in this version using the same logic/event handling as the existing functions.

[AudioPick_v0.3.8.zip](https://github.com/rain-fighters/AudioPick/files/13763876/AudioPick_v0.3.8.zip)

Tests OK on all resource tests using direct media controls and javascript triggers. Tests OK on Spotify, Soundcloud, Youtube, and Reddit. Let me know if I missed anything!

**Comment by necropola** [comment]
Yeah, looks a lot better now. Still some minor issues, but I'll push 0.3.8 after some further tests and some more white space/indention cleanup, i. e. you need to base your next version (I'm sure there will be one 😉) on the upcoming push to GitHub.

Meanwhile, thanks a **LOT** for your work.  The fact that **AudioPick** now supports **Spotify** and **Souncloud**, i. e. sites which do not insert there Media Elements into the DOM tree is fantastic. Not to mention all the changes that were needed to adjust the extension for Manifest V3. ⭐ ⭐ ⭐

**Comment by necropola** [comment]
@XanSama I'ved push 0.3.8 + some mroe white space cleanup and minor adjustments.

One thing that still does not (reliabley) work is opening a SoundCloud link in a new tab,  e. g.
>  https://soundcloud.com/arenanet/mystery-music?in=arenanet/sets/guild-wars-2-season-2-ep-1&utm_source=clipboard&utm_medium=text&utm_campaign=social_sharing

which then automatically starts to play and apparently does not inject the listener. Interestingly it seems to work, when you have stored a domain or global device other than "default". A timing issue?

I was also wondering about this: https://github.com/rain-fighters/AudioPick/blob/eb8d2580e90e073e62f17ac4db5182700fada02e/extension/main.js#L23-L26
- Why is this even needed and
- shouldn't it be testing for  `e.detail === "default"`?

I mean, we are setting `window.activeSinkId` before dispatching the message, but the less dependancies from global variables the better, right?

Cheers

**Comment by XanSama** [comment]
That check is needed because AudioContext objects don't support "default" as a sinkId. If you want an AudioContext to use the default sinkId you need to submit a blank value instead. The SoundCloud issue you describe may be related to that; I realize I didn't put the same logic in the code that sets the initial sinkId.

You're right that we should be using e.detail there; missed that one in my haste.

I've just done some quick testing and it seems like all objects will accept a blank sinkId and interpret it as default so I've removed the object type check and now just set the sinkId to "" if the selected device is "default" regardless. I've also added the logic to the code that sets the initial value (just below the lines you quoted).

[main.zip](https://github.com/rain-fighters/AudioPick/files/13767252/main.zip)

**Comment by necropola** [comment]
@XanSama Thansk a lot!

I still (even with the new `main.js`) occasionally manage to get **SoundCloud** to not work, i. e. it appears that the listener did not get injected. Even not when stopping and starting audio again using the controls at the bottom. Ususally reloading the page helps. I wonder, how they trigger (auto)`play` without **AudioPick** being able to inject the listener, i. e. how they sometimes manage to set everything up (create all detached Audio elements), before the extension is able to inject itself into any prototype methods.  

The other thing I'm still not (fully) convinced of is that we actually need to handle **shadowRoots**.  IMHO we either get those elements via injecting into `prototype.play` (while they are still detached) or via the **Observer** (when added to the DOM tree and becoming visible).
  
Anyway, I'm **VERY HAPPY** about the current state/functionality of the **Manifest V3 Version**.  Hence I'm heading towards releasing it on **GitHub** and then on the **Chrome Webstore**, i. e. 

- Check for / fix typos in comments
- Adjust code style, e. g. placement of braces (if you don't mind)
- Prettify the `popup.html` (inculding a new icon)
- Implement/Update the new `rain-fighters/AudioPick` web site (using **hugo**)
- Update the documentation on **GitHub** (`README.md`, `CHANGELOG.md`) 
- Merge the `manifest-v3` branch back into `master/main`
- Release on **GitHub**
- Update the information on the **Chrome Webstore**
- Release on the **Chrome Webstore**

Thanks again for this 🎄**WONDERFUL CHRISTMAS PRESENT** 🎄

**Comment by necropola** [comment]
@XanSama I'm thinking about removing the option to set/store a global audio sink/device to reduce the number of unneeded/unwanted microphone access exceptions added by the extension.

Since we are now able to set per domain sinks/devices (which we could not before), I think we would loose very littlle functionality, but win a lot of user acceptance.  

This also better matches the fact that we do not have a global options panel and that you can only open the popup for HTTPS tabs/sites.
  
What do you think?

**Comment by XanSama** [comment]
> I still (even with the new main.js) occasionally manage to get SoundCloud to not work, i. e. it appears that the listener did not get injected. Even not when stopping and starting audio again using the controls at the bottom. Ususally reloading the page helps. I wonder, how they trigger (auto)play without AudioPick being able to inject the listener, i. e. how they sometimes manage to set everything up (create all detached Audio elements and start the audio), before the extension is able to inject itself into any prototype methods.

I'm unable to replicate this on my end. If it is a race condition I wonder if it could be mitigated by adjusting the load order of the scripts in manifest.json so main.js loads before content.js? Might be worth a try.

> The other thing I'm still not (fully) convinced of is that we actually need to handle shadowRoots. IMHO we either get those elements via injecting into prototype.play (while they are still detached) or via the Observer (when added to the DOM tree and becoming visible).

The issue with shadowRoots is that if they're created with the mode:closed option the observer won't catch them and our code in main.js wont be able to access them or their children after creation. If we don't find them with the observer or initial document parsing and they don't call .play() then we're SoL. By hooking attachShadow and attaching the observer to the shadowRoot's handle at the time of creation we avoid being locked out.

There's actually a case the current code doesn't address: if the page has created a closed shadowRoot -before- our code loads that shadowRoot and its children are invisible to us. We could address this using the chrome.dom.openOrClosedShadowRoot API from content.js for our queryShadowSelectorAll function, but then we'd need to move/copy some of the observer & sinkId/event code into content.js to get value from it.

If we wanted to do that the function would be something like this:

```
function queryShadowSelectorAll(selectors, e = document) {
	// Use chrome.dom API to find even pre-existing "closed" shadowRoots.
	const getShadowRoot = chrome.dom.openOrClosedShadowRoot;
	// Recursive function to locate all shadowRoots for the provided element.
	const getShadowRoots = function (el) {
		return Array.from(el.querySelectorAll("*"))
			.filter((ce) => Boolean(getShadowRoot(ce)))
				.flatMap((ce) => [
					getShadowRoot(ce),
					...getShadowRoots(getShadowRoot(ce))
				]);
	};
	var results = [];
	// Get every element from every shadowRoot.
	getShadowRoots(e).forEach(function (s) {
		s.querySelectorAll(":host *").forEach(
			function (el) {
				results.push(el);
			}
		);
	});
	// Filter results to our selectors.
	return results.filter(
		function (r) {
			return r.matches(selectors);
		}
	);
};
```

> Adjust code style, e. g. placement of braces (if you don't mind)

I don't mind. Originally I had it formatted to please jslint but you should format it however you want to maintain it.

> I'm thinking about removing the option to set/store a global audio sink/device to reduce the number of unneeded/unwanted microphone access exceptions added by the extension.
Since we are now able to set/store per domain sinks/devices (which we could not before), we would loose very littlle functionality, but win a lot of user acceptance.

That's fair. It does seem unlikely that someone would have a default audio device set for their workstation and want to use a different default for all browsing. I have no strong feelings one way or another on this change.

> This also better matches the fact that we do not have a global options panel and that you can only open the popup for HTTPS tabs/sites.

I'm not sure I'm a fan of blanket disabling http coverage. I guess it's a layer of security against a targeted MITM from someone who knew the extension was in use but that seems... unlikely. Definitely makes testing less convenient if you're not running SSL on your dev server. I'll defer to you on this as the maintainer/publisher of the version in the Chrome store.

Regarding not having an options panel: I was just pondering if we should maybe make one to at least be able to see and remove stored domain preferences. Can't list the devices easily from there but displaying stored data should be doable.

I've also been wondering if it's maybe worthwhile to implement something similar to the sinkId change we're doing but for input devices. Not to actually change input devices used by the websites (though we could, I suppose) but to force them to use a non-existent sink or otherwise disable actual access to the microphone even though we granted the permission...

**Comment by necropola** [comment]
> I'm unable to replicate this on my end. If it is a race condition I wonder if it could be mitigated by adjusting the load order of the scripts in manifest.json so main.js loads before content.js? Might be worth a try.

I'll give it a try, i. e. change the order in `manifest.json`.

> The issue with shadowRoots is that if they're created with the mode:closed option the observer won't catch them and our code in main.js wont be able to access them or their children after creation. If we don't find them with the observer or initial document parsing and they don't call .play() then we're SoL. By hooking attachShadow and attaching the observer to the shadowRoot's handle at the time of creation we avoid being locked out.

I think I'll just leave the **shadowRoot** handling as it is at the moment and wait, until we bump into actual cases ***in the wild*** which require further actions.

> I'm not sure I'm a fan of blanket disabling http coverage

I thought that `.setSinkId()` requires a secure context anyway (https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/setSinkId)? Does chrome allow this in on unencrytped pages, too?

I'm planning to provide our test page with some additional demo links to YouTube (Music), Spotify, SoundCloud, ...  on the new (SSL encrypted) `rain-fighters/AudioPick` web site to provide a ***Reference of Test(ed) Scenarios***. 

> Regarding not having an options panel: I was just pondering if we should maybe make one to at least be able to see and remove stored domain preferences. Can't list the devices easily from there but displaying stored data should be doable.

Assuming that we are currently storing **Device Names** (possibly edited by the user)  and not **Device Ids** (automatically assigned by the system), I was already thinking about lettting the popup ***merge-match*** the stored entries with the actually found device names (grey/disabled row for an unmatched stored value) and have a separate column/marker to indicate which device is currently stored as the default for that domain.

> I've also been wondering if it's maybe worthwhile to implement something similar to the sinkId change we're doing but for input devices. Not to actually change input devices used by the websites (though we could, I suppose) but to force them to use a non-existent sink or otherwise disable actual access to the microphone even though we granted the permission...

Hmm, I'lll give it a thought. For now I think we should focus on minimizing the number of sites we are creating microphone exceptions for, i. e. to only those for which the user actually picked or stored a non-default device . And set it back to "ask" when the user changes it back to system default (as we already do, I think).

**Comment by necropola** [comment]
## 2d / 2colors?
 ![APV3_Icon-580x580-source](https://github.com/rain-fighters/AudioPick/assets/7461137/240314ef-a347-4a41-bf75-a05620fae031)

## 3d / 2 colors?
2. ![APV3_Icon-512x512-dark-sphere](https://github.com/rain-fighters/AudioPick/assets/7461137/eaa1c5bd-838f-4db9-b498-05d2ef687203)

## ... or 3d / 3 colors
![APV3_Icon512](https://github.com/rain-fighters/AudioPick/assets/7461137/422757a7-9372-4d6a-8c48-2ca7cf756c68)

**Comment by XanSama** [comment]
> I thought that .setSinkId() requires a secure context anyway (https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/setSinkId)? Does chrome allow this in on unencrytped pages, too?

Chrome and Edge both seem to allow it on HTTP with the <all_urls> setting.

> Assuming that we are currently storing Device Names (possibly edited by the user) and not Device Ids (automatically assigned by the system), I was already thinking about lettting the popup merge-match the stored entries with the actually found device names (grey/disabled row for an unmatched stored value) and have a separate column/marker to indicate which device is currently stored as the default for that domain.

That's an excellent idea.

> Hmm, I'lll give it a thought. For now I think we should focus on minimizing the number of sites we are creating microphone exceptions for, i. e. to only those for which the user actually picked or stored a non-default device . And set it back to "ask" when the user changes it back to system default (as we already do, I think).

Agreed. Currently if we manually select default from the pop-up it does set it back to ask. I was originally hoping for a more robust logic than that but didn't manage to implement anything better.

Related to both points I think it would be logical for setting a domain back to the default device to clear the stored value for that domain.

> 2d / 2 colors?

I like this one best. Still looks good at 16x16 for icon sizing, too.

**Comment by necropola** [comment]
@XanSama As you can see, I've pushed a few changes ...

Swapping the order of injected scripts did not solve my issue with not intercepting **SoundCloud** early enough when a new tab starts to autoplay immediately. But since a reload usually fixes it, it's not my top priority at the moment.

I have removed the option to store a global **`preferred`**  (new term to avoid confusion with **`default`**) device, but I think we are still requesting microphone access too often. Probably an issue with `undefined != "" != "default"` and/or not correctly checking whether anything has changed and the microphone permission is actually required.

The merge/matching idea for the popup's device list is still on my table. Maybe I'll also find the reason for thos eunneeded microphone permissions while I'm on it. 

Cheers

**Comment by XanSama** [comment]
Looking good :)

At a glance I can't see why we'd be requesting mic permissions when they're not needed but admittedly I'm wiped from Christmas and have picked up a case of COVID so my brain isn't performing at peak.

Also I think I'm going to step back and let you run with it from here. I don't even actually use the extension :sweat_smile: I just picked it up because I needed an example of how to override sink IDs for a private project. Figured making it work again would be a good crash course in the related API.

That is to say... Thank you for all of your work! AudioPick has been an invaluable learning resource for me.

Happy New Year and good luck going forward :)

**Comment by necropola** [comment]
It was a lot of fun working with you. Glad you got something out of it, too.
Maybe we meet again another time, maybe not ... You never know ...

Best wishes for the future!, Thanks a LOT. Happy New Year and Get Better Soon!

**Comment by necropola** [comment]
A small change that makes a big difference: 77167ab

**Comment by necropola** [comment]
Looks like we should prefix the names of global variables (and functions) we insert into MAIN

![image](https://github.com/rain-fighters/AudioPick/assets/7461137/7ad45e46-941e-4152-bd08-b97934d7f0f1)

**Comment by necropola** [comment]
Another issue/error may occur when  the extension has just been (re-)loaded,
but the active tab has not been refreshed  i .e  the (new version of the) content script is not yet available to respond to messages from the popup.

![image](https://github.com/rain-fighters/AudioPick/assets/7461137/6e1829a3-578f-4a94-8464-b9f30bb5ccea)

We need to catch the error here:

https://github.com/rain-fighters/AudioPick/blob/77167abf6663a2f173c2167d6cb8e721ddc28c67/extension/popup.js#L174-L180

**Comment by necropola** [comment]
I've just pushed a major popup / UI Enhancement 1db7a93 ... getting close to release. 

🧨  **HAPPY NEW YEAR EVERYONE** 🧨

**Comment by necropola** [comment]
I've just released version `0.3.8` on **GitHub** and uploaded it to the **Chrome WebStore** for review. Let's see how long that takes ...

**Comment by necropola** [comment]
Version `0.3.8`  has been successfully reviewed by **Google**, but since I found a few issues, I've decided to not publish it on the **Chrome WebStore**. Instead I fixed those issues, released  `0.3.9` on **GitHub** and uploaded the new version to the **WebStore** for review.

**Comment by necropola** [comment]
Version `0.3.9` has been published on the **Chrome WebStore**.

Thanks again to @XanSama for rewriting the old Manifest V2 extension for Manifest V3 and making it work again. ❤️

Closing this issue ...
